pkg/controllers/vpc/vpc_controller.go

```diff
@@ -58,51 +57,74 @@ func (r *VPCReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.R
-				log.Error(err, "add finalizer", "vpc", req.NamespacedName)
+				log.Error(err, "add finalizer", "VPC", req.NamespacedName)
 				updateFail(r.Service.NSXConfig, &ctx, obj, &err, r.Client)
 				return common.ResultRequeue, err
 			}
-			log.V(1).Info("added finalizer on VPC CR", "vpc", req.NamespacedName)
+			log.V(1).Info("added finalizer on VPC CR", "VPC", req.NamespacedName)
 		}
 
-		cVpc, err := r.Service.CreateorUpdateVPC(obj)
+		createdVpc, err := r.Service.CreateorUpdateVPC(obj)
 		if err != nil {
-			log.Error(err, "operate failed, would retry exponentially", "vpc", req.NamespacedName)
+			log.Error(err, "operate failed, would retry exponentially", "VPC", req.NamespacedName)
 			updateFail(r.Service.NSXConfig, &ctx, obj, &err, r.Client)
-			return common.ResultRequeue, err
+			return common.ResultRequeueAfter10sec, err
 		}
-		updateSuccess(r.Service.NSXConfig, &ctx, obj, r.Client, *cVpc.Path, "")
+
+		snatIP, path, cidr := "", "", ""
+		// currently, auto snat is not exposed, and use default value True
+		// checking autosnat to support future extension in vpc configuration
+		if *createdVpc.ServiceGateway.AutoSnat {
+			snatIP, err = r.Service.GetDefaultSNATIP(*createdVpc)
+			if err != nil {
+				log.Error(err, "failed to read default SNAT ip from VPC", "VPC", createdVpc.Id)
+				return common.ResultRequeueAfter10sec, err
+			}
+		}
+
+		// if lb vpc enabled, read avi subnet path and cidr
+		// nsx bug, if set LoadBalancerVpcEndpoint.Enabled to false, when read this vpc back,
+		// LoadBalancerVpcEndpoint.Enabled will become a nil pointer.
+		if createdVpc.LoadBalancerVpcEndpoint.Enabled != nil && *createdVpc.LoadBalancerVpcEndpoint.Enabled {
+			path, cidr, err = r.Service.GetAVISubnetInfo(*createdVpc)
+			if err != nil {
+				log.Error(err, "failed to read lb subnet path and cidr", "VPC", createdVpc.Id)
+				return common.ResultRequeueAfter10sec, err
+			}
+		}
+
+		updateSuccess(r.Service.NSXConfig, &ctx, obj, r.Client, *createdVpc.Path, snatIP, path, cidr)
 	} else {
 		if controllerutil.ContainsFinalizer(obj, commonservice.VPCFinalizerName) {
 			metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteTotal, common.MetricResTypeVPC)
 			vpcs := r.Service.GetVPCsByNamespace(obj.GetNamespace())
+			// if nsx resource do not exist, continue to remove finalizer, or the crd can not be removed
 			if len(vpcs) == 0 {
 				// when nsx vpc not found in vpc store, do not retry
-				notFound := errors.New("can not find vpc from vpc stroe")
-				log.Error(notFound, "can not find vpc from vpc store. put event back to queue")
-				return common.ResultNormal, nil
-			}
-			vpc := vpcs[0]
-			if err := r.Service.DeleteVPC(*vpc.Path); err != nil {
-				log.Error(err, "delete failed, would retry exponentially", "vpc", req.NamespacedName)
-				deleteFail(r.Service.NSXConfig, &ctx, obj, &err, r.Client)
-				return common.ResultRequeueAfter10sec, err
-			}
+				log.Info("can not find VPC in store. put event back to queue")
```

Comment:
looks like it is not re-enqueue?

Reply:
yes, no need to requeue here, if no nsx vpc, we can directly remove finalizer and remove the vpc cr. this is a bugfix i found in rainy case. i manually delete nsx vpc on nsx ui, and restart operator, it will block vpc cr delete event and loop forever if we put it back to queue. changed the log message and comments.